v3/src/components/graph/components/parent-toggles.tsx

```diff
@@ -0,0 +1,242 @@
+  const caseIDs = dataset?.cases.map((c) => c.__id__) ?? []
+  const caseButtons = createCaseButtons({ caseIDs, dataset, isCollectionSet, hiddenCases })
+  const caseButtonsListWidth = caseButtons.reduce((acc, button) => acc + button.width + TEXT_OFFSET, 0)
+  const [isOnlyLastShown, setIsOnlyLastShown] = useState(false)
+  const firstVisibleIndex = useRef(0)
+  const lastVisibleIndex = useRef(caseButtons.length - 1)
+  const toggleButtonText = hiddenCases.length > 0 ? t("DG.NumberToggleView.showAll") : t("DG.NumberToggleView.hideAll")
+  const toggleTextWidth = measureText(toggleButtonText, BUTTON_FONT)
+  const lastCheckbox = isOnlyLastShown ? t("DG.NumberToggleView.lastChecked") : t("DG.NumberToggleView.lastUnchecked")
+  const lastButtonText = `${lastCheckbox} ${t("DG.NumberToggleView.lastLabel")}`
+  const lastButtonWidth = measureText(lastButtonText, BUTTON_FONT)
+  const [showLeftButton, setShowLeftButton] = useState(false)
+  const [showRightButton, setShowRightButton] = useState(false)
+  const [buttonsListOffset, setButtonsListOffset] = useState(0)
+  const [buttonContainerWidth, setButtonContainerWidth] = useState(0)
+
+  // Returns the currently available width for the button list, the total width of the buttons that can fit within
+  // that available width from the given `firstIndex`, and the index of the last visible button.
+  const buttonContainerDetails = useCallback((firstIndex: number) => {
+    let buttonsVisibleWidth = 0
+    let lastIndex = 0
+    const usedWidth = toggleTextWidth + TEXT_OFFSET + lastButtonWidth + TEXT_OFFSET + SCROLL_BUTTON_WIDTH * 2
+    const availableWidth = tileWidth - usedWidth
+
+    for (let i = firstIndex; i < caseButtons.length; i++) {
+      buttonsVisibleWidth += caseButtons[i].width + TEXT_OFFSET
+      if (buttonsVisibleWidth > availableWidth) {
+        lastIndex = i - 1
+        buttonsVisibleWidth -= caseButtons[i].width + TEXT_OFFSET
+        break
+      }
+      lastIndex = i
+    }
+
+    return { availableWidth, buttonsVisibleWidth, lastIndex }
+  }, [caseButtons, lastButtonWidth, tileWidth, toggleTextWidth])
+
+  useEffect(function init() {
+    const { availableWidth, buttonsVisibleWidth, lastIndex } = buttonContainerDetails(firstVisibleIndex.current)
+    setButtonContainerWidth(buttonsVisibleWidth)
+    if (caseButtonsListWidth > availableWidth) {
+      lastVisibleIndex.current = lastIndex
+      setShowRightButton(true)
+      setShowLeftButton(true)
+    } else {
+      setShowRightButton(false)
+      setShowLeftButton(false)
+    }
+  // Only run this once to initialize the button container width, last visible index, and scroll button visibility.
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [])
+
+  const handleToggleAll = () => {
+    if (hiddenCases.length > 0) {
+      dataConfig?.clearHiddenCases()
+      setIsOnlyLastShown(false)
+    } else {
+      dataConfig?.setHiddenCases(Array.from(dataConfig.allCaseIDs))
```

Comment:
@kswenson Yes, I think you're right, although I don't know that anything other than activating the feature should actually be undoable. If I understand correctly, in V2 clicking any of the toggle buttons, the Last button, or the Show/Hide All buttons is not undoable. Is that how it should be in V3, @bfinzer ?

Reply:
The `hiddenCases` are part of the model, so I would say changes to them should be undoable unless there's a compelling reason to do otherwise. Just because v2 did it that way doesn't mean v3 needs to do the same.